File: shared/src/commonMain/kotlin/io/github/thatsfguy/reticulum/transport/Hdlc.kt
package io.github.thatsfguy.reticulum.transport

/**
 * HDLC frame encode/decode for Reticulum's TCP interface.
 *
 * rnsd's TCPClientInterface / TCPServerInterface frames every raw
 * Reticulum packet with HDLC before writing it to the socket. Wire
 * format is:
 *
 *   FLAG (0x7E) || escaped(packet_bytes) || FLAG (0x7E)
 *
 * Escape replaces in-band 0x7D with 0x7D 0x5D, and 0x7E with 0x7D 0x5E.
 *
 * Same escape logic as KISS but different byte values, and unlike KISS
 * there is no command byte prefix and no RSSI/SNR sidecar — the frame
 * IS the Reticulum packet.
 *
 * Source of truth: RNS/Interfaces/TCPInterface.py class HDLC.
 * Port of reference/js-reference/hdlc.js.
 */

const val HDLC_FLAG     = 0x7E
const val HDLC_ESC      = 0x7D
const val HDLC_ESC_MASK = 0x20

private val HDLC_FLAG_B = HDLC_FLAG.toByte()
private val HDLC_ESC_B  = HDLC_ESC.toByte()

/** Wrap one complete Reticulum packet into an HDLC frame. */
fun buildHdlcFrame(data: ByteArray): ByteArray {
    // Worst case: every byte needs escaping → 2x size, plus two flags.
    val out = ArrayList<Byte>(data.size + 2 + (data.size shr 3))
    out.add(HDLC_FLAG_B)
    for (b in data) {
        val v = b.toInt() and 0xFF
        if (v == HDLC_FLAG || v == HDLC_ESC) {
            out.add(HDLC_ESC_B)
            out.add((v xor HDLC_ESC_MASK).toByte())
        } else {
            out.add(b)
        }
    }
    out.add(HDLC_FLAG_B)
    return out.toByteArray()
}

/**
 * Streaming HDLC parser. Feed bytes as they arrive from the socket.
 * Partial frames are buffered across feeds, so chunk boundaries do
 * not matter. Empty frames (two FLAGs back to back) are silently
 * dropped — rnsd uses FLAG as both delimiter and keepalive idle.
 */
class HdlcParser(private val onFrame: (data: ByteArray) -> Unit) {
    private val buf = ArrayList<Byte>(512)
    private var inFrame = false
    private var escape = false

    /** Hard cap on a single in-flight frame. Without this a peer that
     *  emits a FLAG and then withholds the closing FLAG could grow the
     *  buffer without bound. RNS packets are at most a few hundred bytes
     *  + framing, so 64 KB is comfortably above legitimate traffic. */
    private val maxFrameBytes = 64 * 1024

    fun feed(bytes: ByteArray) {
        for (raw in bytes) {
            val b = raw.toInt() and 0xFF
            if (b == HDLC_FLAG) {
                if (inFrame && buf.isNotEmpty()) {
                    onFrame(buf.toByteArray())
                }
                buf.clear()
                inFrame = true
                escape = false
                continue
            }
            if (!inFrame) continue

            if (escape) {
                escape = false
                buf.add((b xor HDLC_ESC_MASK).toByte())
            } else if (b == HDLC_ESC) {
                escape = true
            } else {
                buf.add(raw)
            }

            if (buf.size > maxFrameBytes) {
                // Runaway frame — discard and resync on the next FLAG.
                buf.clear()
                inFrame = false
                escape = false
            }
        }
    }

    fun reset() {
        buf.clear()
        inFrame = false
        escape = false
    }
}

File: reference/js-reference/hdlc.js
// js/hdlc.js — HDLC frame encode/decode for Reticulum's TCP interface.
//
// rnsd's TCPClientInterface / TCPServerInterface frames every raw
// Reticulum packet with HDLC before writing it to the socket, and
// unframes it on the receive side. Wire format is:
//
//   FLAG (0x7E) || escaped(packet_bytes) || FLAG (0x7E)
//
// where escaping replaces any in-band 0x7D with 0x7D 0x5D and any
// in-band 0x7E with 0x7D 0x5E. The reverse applies on unescape.
//
// This is a separate module from kiss.js because:
//   * framing bytes are different (0x7E / 0x7D vs 0xC0 / 0xDB)
//   * HDLC has no command byte prefix — the frame IS the packet
//   * there is no RSSI/SNR metadata framing like KISS has
//
// Source: RNS/Interfaces/TCPInterface.py class HDLC.

'use strict';

export const FLAG     = 0x7E;
export const ESC      = 0x7D;
export const ESC_MASK = 0x20;

// Wrap one complete Reticulum packet into an HDLC frame. Returns
// FLAG || escaped(data) || FLAG.
export function encodeFrame(data) {
  const out = [FLAG];
  for (const b of data) {
    if (b === ESC || b === FLAG) {
      out.push(ESC, b ^ ESC_MASK);
    } else {
      out.push(b);
    }
  }
  out.push(FLAG);
  return new Uint8Array(out);
}

// Streaming HDLC parser. Feed it bytes as they arrive on the
// transport (WebSocket messages, TCP chunks, whatever) and it calls
// onFrame(bytes) for each complete frame boundary it sees. Buffers
// partial frames across feeds so input chunk sizes do not matter.
export class HdlcParser {
  constructor(onFrame) {
    this.onFrame = onFrame;
    this._buf = [];
    this._inFrame = false;
    this._escape = false;
  }

  feed(bytes) {
    for (const b of bytes) {
      if (b === FLAG) {
        // A FLAG terminates the current frame (if any) and starts
        // the next. Empty frames (two FLAGs back to back) are
        // silently dropped — rnsd uses FLAG as both delimiter and
        // keepalive.
        if (this._inFrame && this._buf.length > 0) {
          this.onFrame(new Uint8Array(this._buf));
        }
        this._buf = [];
        this._inFrame = true;
        this._escape = false;
        continue;
      }

      if (!this._inFrame) continue;

      if (this._escape) {
        this._escape = false;
        this._buf.push(b ^ ESC_MASK);
      } else if (b === ESC) {
        this._escape = true;
      } else {
        this._buf.push(b);
      }
    }
  }

  reset() {
    this._buf = [];
    this._inFrame = false;
    this._escape = false;
  }
}
